rclpy/rclpy/handle.py

```diff
@@ -117,13 +147,48 @@ def __exit__(self, type_, value, traceback):
         self._return_capsule()
 
     def __destroy(self):
-        # Assume lock is held
-        assert not self.__lock.acquire(blocking=False)
-        # Assume capsule has not been destroyed
-        assert self.__valid
```

Comment:
Destruction can be delayed waiting for dependents. Previously `self.__valid = False` meant the capsule itself has been destroyed. Now setting it to false means `destroy()` was called and `__use_count` hit zero. The purpose is to prevent new users of this handle while waiting for dependent handles to be destroyed.